File: sdk/ios/Sources/DuallerSDK/PageWebViewPool.swift
import UIKit
import WebKit

/// LRU-based WebView pool for iOS page reuse.
public final class PageWebViewPool {
    private let context: UIApplication
    private let config: EngineConfig
    private var maxCacheSize: Int

    private var idleMap = [String: DuallerWebView]()
    private var accessOrder: [String] = []
    private var allWebViews = [String: DuallerWebView]()

    init(context: UIApplication, config: EngineConfig, maxCacheSize: Int = 5) {
        self.context = context
        self.config = config
        self.maxCacheSize = maxCacheSize
    }

    /// Acquire a WebView for the given route. Creates one if none idle.
    public func acquire(route: String) throws -> DuallerWebView {
        // Check idle pool first.
        if let idle = idleMap[route] {
            // Move to front of access order (most recently used).
            accessOrder.removeAll { $0 == route }
            accessOrder.append(route)
            print("[PageWebViewPool] Reusing idle WebView for \(route)")
            return idle
        }

        // Evict LRU if at capacity.
        while idleMap.count >= maxCacheSize && !accessOrder.isEmpty {
            destroyOldest()
        }

        print("[PageWebViewPool] Creating new WebView for \(route)")
        let webView = try DuallerWebView(bundleURL: resolvePageURL(route), debugMode: config.debugMode)
        allWebViews[route] = webView
        accessOrder.append(route)
        return webView
    }

    /// Release a WebView back to the pool (hide it, mark idle).
    public func release(_ webView: DuallerWebView, forRoute route: String) {
        webView.isHidden = true
        webView.webView?.stopLoading()
        webView.webView?.pauseTimers()

        idleMap[route] = webView
        print("[PageWebViewPool] Released WebView for \(route) to pool")
    }

    /// Destroy the oldest (least recently used) idle WebView.
    public func destroyOldest() {
        guard let oldestRoute = accessOrder.first else { return }
        accessOrder.removeFirst()
        idleMap.removeValue(forKey: oldestRoute)
        allWebViews.removeValue(forKey: oldestRoute)
        print("[PageWebViewPool] Destroyed oldest WebView for \(oldestRoute) (LRU)")
    }

    /// Clear all WebViews from the pool.
    public func clear() {
        idleMap.removeAll()
        allWebViews.removeAll()
        accessOrder.removeAll()
        print("[PageWebViewPool] Cleared all WebViews from pool")
    }

    public var idleCount: Int { return idleMap.count }

    /// Resolve a page route to its bundle URL.
    private func resolvePageURL(_ route: String) -> URL {
        let pageName = route.components(separatedBy: "/").last ?? route
        let basePath = config.bundleURL.deletingLastPathComponent().deletingLastPathComponent()
        return basePath.appendingPathComponent("\(route)/\(pageName).js")
    }
}

File: sdk/android/src/main/java/com/dualler/sdk/view/PageWebViewPool.kt
package com.dualler.sdk.view

import android.content.Context
import android.util.Log
import com.dualler.sdk.EngineConfig

/**
 * LRU-based WebView pool for page reuse.
 *
 * When acquire() is called, returns an existing idle WebView for the route
 * or creates a new one. When release() is called, hides and recycles the
 * WebView back to the pool. If the pool exceeds maxCacheSize, the LRU
 * WebView is destroyed.
 */
class PageWebViewPool(private val context: Context, private val config: EngineConfig) {
    private val TAG = "PageWebViewPool"
    private val _idleMap = mutableMapOf<String, DequeItem>()
    private val _accessOrder = mutableListOf<String>()

    var maxCacheSize: Int = 5

    /** All currently managed WebViews (idle + in-use), for cleanup. */
    private val _allWebViews = mutableMapOf<String, DuallerWebView>()

    /** Acquire a WebView for the given route. Creates one if none idle. */
    fun acquire(route: String): DuallerWebView {
        // Check idle pool first.
        val idle = _idleMap[route]
        if (idle != null) {
            // Move to front of access order (most recently used).
            _accessOrder.removeAll { it == route }
            _accessOrder.add(route)
            Log.d(TAG, "Reusing idle WebView for $route")
            return idle.webView
        }

        Log.d(TAG, "Creating new WebView for $route")
        val webView = DuallerWebView(context, config, route)
        _allWebViews[route] = webView
        _accessOrder.add(route)
        return webView
    }

    /** Release a WebView back to the pool (hide it, mark idle). */
    fun release(route: String, webView: DuallerWebView) {
        webView.visibility = android.view.View.INVISIBLE
        webView.pauseTimers()

        _idleMap[route] = DequeItem(webView)
        Log.d(TAG, "Released WebView for $route to pool")
    }

    /** Destroy the oldest (least recently used) idle WebView. */
    fun destroyOldest() {
        if (_accessOrder.isEmpty()) return
        val oldestRoute = _accessOrder.first()
        _accessOrder.remove(oldestRoute)
        val item = _idleMap.remove(oldestRoute)
        item?.webView?.destroy()
        _allWebViews.remove(oldestRoute)
        Log.d(TAG, "Destroyed oldest WebView for $oldestRoute (LRU)")
    }

    /** Get count of idle WebViews in the pool. */
    fun idleCount(): Int = _idleMap.size

    /** Clear all WebViews from the pool. */
    fun clear() {
        for (item in _idleMap.values) {
            item.webView.destroy()
        }
        _idleMap.clear()
        _allWebViews.clear()
        _accessOrder.clear()
        Log.d(TAG, "Cleared all WebViews from pool")
    }

    private data class DequeItem(val webView: DuallerWebView)
}
